File: paper_review/parser.py
"""Deterministic paper parser for .md / .txt sources.

Splits a paper into a title and an ordered list of sections. Section
boundaries are markdown headers (``#`` .. ``######``) or, for plain
text, lines that are exactly a well-known section heading. No network,
no randomness, no timestamps - parsing the same bytes always yields the
same structure.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

SUPPORTED_SUFFIXES = (".md", ".markdown", ".txt")

KNOWN_SECTIONS = (
    "abstract",
    "introduction",
    "background",
    "related work",
    "method",
    "methods",
    "methodology",
    "approach",
    "experimental setup",
    "experiments",
    "evaluation",
    "results",
    "analysis",
    "discussion",
    "conclusion",
    "conclusions",
    "limitations",
    "future work",
    "references",
    "acknowledgements",
    "acknowledgments",
)

_HEADER_RE = re.compile(r"^(#{1,6})\s+(.*\S)\s*$")


@dataclass(frozen=True)
class Section:
    title: str
    level: int
    text: str


@dataclass(frozen=True)
class ParsedPaper:
    title: str
    sections: tuple[Section, ...]
    raw_text: str

    def section_titles(self) -> tuple[str, ...]:
        return tuple(s.title for s in self.sections)

    def full_text(self) -> str:
        return self.raw_text


def parse_paper(path: str | Path) -> ParsedPaper:
    p = Path(path)
    if p.suffix.lower() not in SUPPORTED_SUFFIXES:
        raise ValueError(
            f"unsupported file type {p.suffix!r}; expected one of "
            f"{', '.join(SUPPORTED_SUFFIXES)}"
        )
    return parse_text(p.read_text(encoding="utf-8"))


def parse_text(text: str) -> ParsedPaper:
    norm = text.replace("\r\n", "\n").replace("\r", "\n")
    lines = norm.split("\n")
    title = _detect_title(lines)
    sections = _split_sections(lines)
    return ParsedPaper(title=title, sections=tuple(sections), raw_text=norm)


def _detect_title(lines: list[str]) -> str:
    for ln in lines:
        m = _HEADER_RE.match(ln)
        if m and len(m.group(1)) == 1:
            return m.group(2).strip()
    for ln in lines:
        if ln.strip():
            return ln.strip().lstrip("#").strip()
    return "Untitled"


def _heading_of(line: str) -> tuple[int, str] | None:
    m = _HEADER_RE.match(line)
    if m:
        return len(m.group(1)), m.group(2).strip()
    s = line.strip()
    if s and len(s) <= 60 and s.rstrip(":").strip().lower() in KNOWN_SECTIONS:
        return 2, s.rstrip(":").strip()
    return None


def _split_sections(lines: list[str]) -> list[Section]:
    sections: list[Section] = []
    cur_title: str | None = None
    cur_level = 1
    cur_body: list[str] = []

    def flush() -> None:
        text = "\n".join(cur_body).strip()
        if cur_title is not None:
            sections.append(Section(cur_title, cur_level, text))
        elif text:
            sections.append(Section("Preamble", 1, text))

    for ln in lines:
        heading = _heading_of(ln)
        if heading is not None:
            flush()
            cur_level, cur_title = heading
            cur_body = []
        else:
            cur_body.append(ln)
    flush()
    return sections


__all__ = [
    "KNOWN_SECTIONS",
    "ParsedPaper",
    "Section",
    "SUPPORTED_SUFFIXES",
    "parse_paper",
    "parse_text",
]

File: web/internal/review/parser.go
package review

import (
	"regexp"
	"strings"
)

var headerRe = regexp.MustCompile(`^(#{1,6})\s+(.*\S)\s*$`)

var knownSections = map[string]bool{
	"abstract": true, "introduction": true, "background": true,
	"related work": true, "method": true, "methods": true,
	"methodology": true, "approach": true, "experimental setup": true,
	"experiments": true, "evaluation": true, "results": true,
	"analysis": true, "discussion": true, "conclusion": true,
	"conclusions": true, "limitations": true, "future work": true,
	"references": true, "acknowledgements": true, "acknowledgments": true,
}

// ParseText parses raw paper text into a ParsedPaper.
func ParseText(text string) ParsedPaper {
	norm := strings.ReplaceAll(text, "\r\n", "\n")
	norm = strings.ReplaceAll(norm, "\r", "\n")
	lines := strings.Split(norm, "\n")
	return ParsedPaper{
		Title:    detectTitle(lines),
		Sections: splitSections(lines),
		RawText:  norm,
	}
}

func detectTitle(lines []string) string {
	for _, ln := range lines {
		if m := headerRe.FindStringSubmatch(ln); m != nil && len(m[1]) == 1 {
			return strings.TrimSpace(m[2])
		}
	}
	for _, ln := range lines {
		if strings.TrimSpace(ln) != "" {
			return strings.TrimSpace(strings.TrimLeft(strings.TrimSpace(ln), "#"))
		}
	}
	return "Untitled"
}

func headingOf(line string) (level int, title string, ok bool) {
	if m := headerRe.FindStringSubmatch(line); m != nil {
		return len(m[1]), strings.TrimSpace(m[2]), true
	}
	s := strings.TrimSpace(line)
	if s != "" && len([]rune(s)) <= 60 {
		key := strings.ToLower(strings.TrimSpace(strings.TrimRight(s, ":")))
		if knownSections[key] {
			return 2, strings.TrimSpace(strings.TrimRight(s, ":")), true
		}
	}
	return 0, "", false
}

func splitSections(lines []string) []Section {
	sections := []Section{}
	var curTitle string
	haveTitle := false
	curLevel := 1
	var body []string

	flush := func() {
		text := strings.TrimSpace(strings.Join(body, "\n"))
		if haveTitle {
			sections = append(sections, Section{Title: curTitle, Level: curLevel, Text: text})
		} else if text != "" {
			sections = append(sections, Section{Title: "Preamble", Level: 1, Text: text})
		}
	}

	for _, ln := range lines {
		if level, title, ok := headingOf(ln); ok {
			flush()
			curLevel, curTitle, haveTitle = level, title, true
			body = nil
		} else {
			body = append(body, ln)
		}
	}
	flush()
	return sections
}

// SectionTitles returns the ordered section titles.
func (p ParsedPaper) SectionTitles() []string {
	out := make([]string, len(p.Sections))
	for i, s := range p.Sections {
		out[i] = s.Title
	}
	return out
}
